l1-contracts/src/shared/libraries/BN254Lib.sol

```diff
@@ -0,0 +1,317 @@
+  ///      Reverts if the point is not on‐curve or the scalar ≥ p.
+  function g1Mul(G1Point memory p, uint256 s) internal view returns (G1Point memory output) {
+    uint256[3] memory input;
+    input[0] = p.x;
+    input[1] = p.y;
+    input[2] = s;
+
+    bool success;
+    assembly {
+      success :=
+        staticcall(
+          sub(gas(), 2000),
+          0x07, // precompile address
+          input,
+          0x60, // input size = 3 × 32 bytes
+          output,
+          0x40 // output size = 2 × 32 bytes
+        )
+    }
+    if (!success) revert mulPointFail();
+    return output;
+  }
+
+  function bn254Pairing(
+    G1Point memory g1a,
+    G2Point memory g2a,
+    G1Point memory g1b,
+    G2Point memory g2b
+  ) internal view returns (bool) {
+    uint256[12] memory input;
+
+    input[0] = g1a.x;
+    input[1] = g1a.y;
+    input[2] = g2a.x1;
+    input[3] = g2a.x0;
+    input[4] = g2a.y1;
+    input[5] = g2a.y0;
+
+    input[6] = g1b.x;
+    input[7] = g1b.y;
+    input[8] = g2b.x1;
+    input[9] = g2b.x0;
+    input[10] = g2b.y1;
+    input[11] = g2b.y0;
+
+    uint256[1] memory result;
+    bool didCallSucceed;
+    assembly {
+      didCallSucceed :=
+        staticcall(
+          sub(gas(), 2000),
+          8,
+          input,
+          0x180, // input size = 12 * 32 bytes
+          result,
+          0x20 // output size = 32 bytes
+        )
+    }
+    require(didCallSucceed, pairingFail());
+    return result[0] == 1;
+  }
+
+  function hashToPoint(bytes32 domain, bytes memory message)
+    internal
+    view
+    returns (G1Point memory output)
+  {
+    bytes32 hashed = keccak256(abi.encode(domain, message));
+    uint256 x = uint256(hashed) % BASE_FIELD_ORDER;
+    uint256 y;
+    bool found = false;
+    while (true) {
+      y = mulmod(x, x, BASE_FIELD_ORDER);
+      y = mulmod(y, x, BASE_FIELD_ORDER);
+      y = addmod(y, 3, BASE_FIELD_ORDER);
+      (y, found) = sqrt(y);
+      if (found) {
+        output = G1Point({x: x, y: y});
+        break;
+      }
+      x = addmod(x, 1, BASE_FIELD_ORDER);
+    }
+    require(found, noPointFound());
+    return output;
+  }
+
+  function sqrt(uint256 xx) internal view returns (uint256 x, bool hasRoot) {
+    bool callSuccess;
+    assembly {
+      let freeMem := mload(0x40)
+      mstore(freeMem, 0x20)
+      mstore(add(freeMem, 0x20), 0x20)
+      mstore(add(freeMem, 0x40), 0x20)
+      mstore(add(freeMem, 0x60), xx)
+      // (N + 1) / 4 = 0xc19139cb84c680a6e14116da060561765e05aa45a1c72a34f082305b61f3f52
+      mstore(add(freeMem, 0x80), 0xc19139cb84c680a6e14116da060561765e05aa45a1c72a34f082305b61f3f52)
+      // N = 0x30644e72e131a029b85045b68181585d97816a916871ca8d3c208c16d87cfd47
+      mstore(add(freeMem, 0xA0), 0x30644e72e131a029b85045b68181585d97816a916871ca8d3c208c16d87cfd47)
+      callSuccess := staticcall(sub(gas(), 2000), 5, freeMem, 0xC0, freeMem, 0x20)
+      x := mload(freeMem)
```

Comment:
this function returns value to freeMem while below returns to 0x00 scratch space